cpp/src/Ice/ConnectionI.cpp

```diff
@@ -2362,7 +2321,11 @@ Ice::ConnectionI::inactivityCheck() noexcept
 {
     // Called by the InactivityTimerTask.
     std::lock_guard lock(_mutex);
-    if (isAtRest())
+    if (_state == StateActive &&
+        _inactivityTimerTaskScheduled &&              // it's false if some other thread canceled the timer while we
+                                                      // were waiting for _mutex
+        !_timer->isScheduled(_inactivityTimerTask))   // make sure this timer task was not rescheduled for later while
```

Comment:
Imagine that while we were waiting for _mutex:
 - thread X canceled the timer task (=> _inactivityTimerTaskScheduled = false)
 - then thread Y scheduled the timer task while sending a heartbeat (=> _inactivityTimerTaskScheduled is true again)

It's a very unlikely "false positive" for _inactivityTimerTaskScheduled.